pkg/operator/ceph/csi/csi-spec.go

```diff
@@ -0,0 +1,249 @@
+	CSIParam.Namespace = namespace
+}
+
+func ValidateCSIParam() error {
+	if EnableRBD {
+		if len(CSIParam.RBDPluginImage) == 0 {
+			return fmt.Errorf("missing csi rbd plugin image")
+		}
+		if len(CSIParam.RegistrarImage) == 0 {
+			return fmt.Errorf("missing csi registrar image")
+		}
+		if len(CSIParam.ProvisionerImage) == 0 {
+			return fmt.Errorf("missing csi provisioner image")
+		}
+		if len(CSIParam.AttacherImage) == 0 {
+			return fmt.Errorf("missing csi attacher image")
+		}
+		if len(CSIParam.RBDPluginSA) == 0 {
+			return fmt.Errorf("missing csi rbd plugin sa")
+		}
+		if len(CSIParam.RBDAttacherSA) == 0 {
+			return fmt.Errorf("missing csi rbd attacher sa")
+		}
+		if len(CSIParam.RBDProvisionerSA) == 0 {
+			return fmt.Errorf("missing rbd provisioner sa")
+		}
+		if len(RBDPluginTemplatePath) == 0 {
+			return fmt.Errorf("missing rbd plugin template path")
+		}
+		if len(RBDProvisionerTemplatePath) == 0 {
+			return fmt.Errorf("missing rbd provisioner template path")
+		}
+		if len(RBDAttacherTemplatePath) == 0 {
+			return fmt.Errorf("missing rbd attacher template path")
+		}
+	}
+
+	if EnableCephFS {
+		if len(CSIParam.CephFSPluginImage) == 0 {
+			return fmt.Errorf("missing csi cephfs plugin image")
+		}
+		if len(CSIParam.RegistrarImage) == 0 {
+			return fmt.Errorf("missing csi registrar image")
+		}
+		if len(CSIParam.ProvisionerImage) == 0 {
+			return fmt.Errorf("missing csi provisioner image")
+		}
+		if len(CSIParam.CephFSPluginSA) == 0 {
+			return fmt.Errorf("missing csi cephfs plugin sa")
+		}
+		if len(CSIParam.CephFSProvisionerSA) == 0 {
+			return fmt.Errorf("missing cephfs provisioner sa")
+		}
+		if len(CephFSPluginTemplatePath) == 0 {
+			return fmt.Errorf("missing cephfs plugin template path")
+		}
+		if len(CephFSProvisionerTemplatePath) == 0 {
+			return fmt.Errorf("missing ceph provisioner template path")
+		}
+	}
+	return nil
+}
+
+func StartCSIDrivers(namespace string, clientset kubernetes.Interface) error {
+	var err error
+
+	if EnableRBD {
+		rbdPlugin, err = templateToDaemonSet("rbdplugin", RBDPluginTemplatePath)
+		if err != nil {
+			return fmt.Errorf("failed to load rbd plugin template: %v", err)
+		}
+		rbdProvisioner, err = templateToStatefulSet("rbd-provisioner", RBDProvisionerTemplatePath)
+		if err != nil {
+			return fmt.Errorf("failed to load rbd provisioner template: %v", err)
+		}
+		rbdAttacher, err = templateToStatefulSet("rbd-attacher", RBDAttacherTemplatePath)
+		if err != nil {
+			return fmt.Errorf("failed to load rbd attacher template: %v", err)
+		}
+	}
+	if EnableCephFS {
+		cephfsPlugin, err = templateToDaemonSet("cephfsplugin", CephFSPluginTemplatePath)
+		if err != nil {
+			return fmt.Errorf("failed to load CephFS plugin template: %v", err)
+		}
+		cephfsProvisioner, err = templateToStatefulSet("cephfs-provisioner", CephFSProvisionerTemplatePath)
+		if err != nil {
+			return fmt.Errorf("failed to load CephFS provisioner template: %v", err)
+		}
+	}
+
+	if rbdPlugin != nil {
```

Comment:
can't we do this after L166 since this is related to RBD?
Also, I believe `rbdPlugin` is always `!= nil` otherwise we would have failed on L165 so this condition might be confusing.

Reply:
my thinking is that there is a good chance only one of rbd and cephfs is deployed. In this case, rbdPlugin can be nil. 

Within rbd or cephfs, there is a also a chance that one of the template might be broken and cause the templating function fail. To prevent such case, it makes sense to defer creating API till all templates are resolved successfully.

Reply:
Makes sense, thanks